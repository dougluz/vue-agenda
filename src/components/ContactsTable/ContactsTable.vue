<script setup lang="ts">
import { useContactStore } from "@/stores/useContactStore";
import ContactAvatar from "../ContactAvatar/ContactAvatar.vue";
import IcEdit from "../icons/IcEdit.vue";
import IcDelete from "../icons/IcDelete.vue";
import { useModalStore } from "@/stores/useModalStore";

const contactsStore = useContactStore();
const modalStore = useModalStore();

const handleDelete = (email: string) => {
  contactsStore.setSelectedContactEmail(email);
  modalStore.setModalStatus("deleteModal", true);
};

// write a function to format a phone number in the format (xx) xxxxx-xxxx
const formatPhoneNumber = (phoneNumber: string) => {
  const phone = phoneNumber.replace(/\D/g, "");
  const phoneLength = phone.length;

  if (phoneLength < 11) {
    return phone.replace(/(\d{2})(\d{4,5})(\d{4})/, "($1) $2-$3");
  }

  return phone.replace(/(\d{2})(\d{1})(\d{4,5})(\d{4})/, "($1) $2 $3-$4");
};
</script>

<template>
  <div class="p-4 w-full">
    <table class="w-full">
      <thead
        class="w-full bg-white p-4 h-10 text-blue-grey border border-light-blue-grey text-left"
      >
        <tr>
          <th class="w-4"></th>
          <th>Contato</th>
          <th>E-mail</th>
          <th>Telefone</th>
          <th class="w-4"></th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="h-10 bg-white px-2 py-4 hover:bg-very-light-pink"
          v-for="contact in contactsStore.contacts"
          :key="contact.email"
        >
          <td class="pl-2 py-4 w-10">
            <ContactAvatar :contactName="contact?.name" />
          </td>
          <td class="py-4">{{ contact?.name }}</td>
          <td class="py-4">{{ contact?.email }}</td>
          <td class="py-4">{{ formatPhoneNumber(contact?.phone) }}</td>
          <td class="flex items-center justify-end gap-6 pr-2 py-4">
            <button><IcEdit /></button>
            <button @click="handleDelete(contact.email!)"><IcDelete /></button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</template>
