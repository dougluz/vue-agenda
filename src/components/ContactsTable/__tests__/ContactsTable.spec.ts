import { describe, it, beforeAll, expect, vi } from "vitest";
import { mount, VueWrapper } from "@vue/test-utils";
import { createTestingPinia } from "@pinia/testing";
import ContactsTable from "../ContactsTable.vue";
import type { Contact } from "@/types/contact";

describe("Contacts Table", () => {
  let wrapper: VueWrapper<any>;

  const contact: Contact = {
    name: "John Doe",
    email: "john@doe.com",
    phone: "1234567890",
  };

  beforeAll(() => {
    wrapper = mount(ContactsTable, {
      global: {
        plugins: [
          createTestingPinia({
            createSpy: vi.fn,
            initialState: {
              contact: { contacts: [contact], getAllContacts: () => [contact] }, // start the counter at 20 instead of 0
            },
          }),
        ],
      },
    });
  });

  it("should match the snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should render the contact data", () => {
    expect(wrapper.text()).toContain(contact.name);
    expect(wrapper.find("tbody").text()).toContain(contact.email);
    expect(wrapper.find("tbody").text()).toContain(contact.phone);
  });
});
