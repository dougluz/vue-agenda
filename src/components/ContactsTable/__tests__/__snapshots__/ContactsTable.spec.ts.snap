// Vitest Snapshot v1

exports[`Contacts Table > should match the snapshot 1`] = `
"<div class=\\"p-4 w-full\\">
  <table class=\\"w-full\\">
    <thead class=\\"w-full bg-white p-4 h-10 text-blue-grey border border-light-blue-grey text-left\\">
      <tr>
        <th class=\\"w-4\\"></th>
        <th>Contato</th>
        <th>E-mail</th>
        <th>Telefone</th>
        <th class=\\"w-4\\"></th>
      </tr>
    </thead>
    <tbody>
      <tr class=\\"h-10 bg-white px-2 py-4 hover:bg-very-light-pink\\">
        <td class=\\"pl-2 py-4 w-10\\"><span class=\\"avatar\\" data-v-689e67aa=\\"\\">J</span></td>
        <td class=\\"py-4\\">John Doe</td>
        <td class=\\"py-4\\">john@doe.com</td>
        <td class=\\"py-4\\">1234567890</td>
        <td class=\\"flex items-center justify-end gap-6 pr-2 py-4\\"><button><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"16\\" height=\\"16\\" viewBox=\\"0 0 16 16\\">
              <g fill=\\"#9198AF\\" fill-rule=\\"nonzero\\">
                <path d=\\"M.15 14.838l-.143.596a.283.283 0 0 0-.006.057L0 15.503a.284.284 0 0 0 .027.115c.014.03.032.06.056.084a.288.288 0 0 0 .268.076l.596-.142-.798-.798zM15.034.752a2.575 2.575 0 0 0-3.637 0L1.093 11.055l-.012.016-.016.022a.278.278 0 0 0-.036.068l-.008.022-.004.008-.712 2.994L1.6 15.481l2.994-.713c.008-.002.015-.006.022-.009a.264.264 0 0 0 .086-.046l.02-.016.008-.005L15.034 4.388a2.574 2.574 0 0 0 0-3.636zM4.902 13.712l.101-.91 6.394-6.393a.285.285 0 1 0-.404-.404l-6.394 6.394-1.364.151.152-1.364L9.78 4.793a.285.285 0 1 0-.404-.404l-6.394 6.393-.91.101 7.91-7.909 2.828 2.829-7.909 7.909zm8.313-8.313L10.387 2.57l.404-.404 2.828 2.829-.404.404zm.808-.808l-2.828-2.829.404-.404 2.829 2.828-.405.405z\\"></path>
              </g>
            </svg></button><button><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"16\\" height=\\"16\\" viewBox=\\"0 0 16 16\\">
              <g fill=\\"#9198AF\\" fill-rule=\\"nonzero\\">
                <path d=\\"M13.881 1.627h-3.808c-.03-.339-.134-.91-.492-1.277A1.147 1.147 0 0 0 8.73 0H6.017c-.34 0-.626.118-.853.35-.357.366-.461.938-.491 1.277H1.407a.271.271 0 1 0 0 .542H13.88a.271.271 0 1 0 0-.542zm-8.663 0c.03-.284.118-.677.336-.9a.614.614 0 0 1 .463-.185h2.712c.19 0 .341.06.463.185.218.223.306.616.336.9h-4.31zM2.48 14.65c.01.467.3 1.35 1.33 1.35h7.668c1.03 0 1.32-.883 1.33-1.345l.497-11.943H1.983L2.48 14.65zm1.91-3.26h6.508a.271.271 0 1 1 0 .542H4.39a.271.271 0 1 1 0-.542zm0 1.356h6.508a.271.271 0 1 1 0 .542H4.39a.271.271 0 1 1 0-.542zm0 1.356h6.508a.271.271 0 1 1 0 .542H4.39a.271.271 0 1 1 0-.542z\\"></path>
              </g>
            </svg></button></td>
      </tr>
    </tbody>
  </table>
</div>"
`;
